examples/pytorch/gat/main.py

```diff
@@ -92,58 +52,76 @@ def train_func(config: Dict):
         optimizer, named_parameters=model.named_parameters()
     )
 
-    address = "localhost:9999"
-    s = Server(address, args.data_dir, 0, len(args.partitions))
-    g = DistributedClient([address])
-    max_id = g.node_count(args.node_type) if args.max_id in [-1, None] else args.max_id
-    dataset = ray.data.range(max_id).repartition(max_id // args.batch_size)
-    pipe = dataset.window(blocks_per_window=4).repeat(args.num_epochs)
+    g = config["get_graph"]()
+    batch_size = 140
+    train_dataset = ray.data.read_text(f"{config['data_dir']}/train.nodes")
+    train_dataset = train_dataset.repartition(train_dataset.count() // batch_size)
+    train_pipe = train_dataset.window(blocks_per_window=4).repeat(config["num_epochs"])
 
     def transform_batch(idx: list) -> dict:
         return model.q.query_training(g, np.array(idx))
 
-    pipe = pipe.map_batches(transform_batch)
+    train_pipe = train_pipe.map_batches(transform_batch)
 
-    for epoch in range(epochs_trained, args.num_epochs):
-        scores = []
-        labels = []
+    test_dataset = ray.data.read_text(f"{config['data_dir']}/test.nodes")
+    test_dataset = test_dataset.repartition(1)
+    test_dataset = test_dataset.map_batches(transform_batch)
+    test_dataset_iter = test_dataset.repeat(config["num_epochs"]).iter_epochs()
+
+    for epoch, epoch_pipe in enumerate(train_pipe.iter_epochs()):
+        model.train()
         losses = []
-        for step, batch in enumerate(dataset):
-            if step < steps_in_epoch_trained:
-                continue
+        for step, batch in enumerate(
+            epoch_pipe.iter_torch_batches(batch_size=batch_size)
+        ):
             loss, score, label = model(batch)
             optimizer.zero_grad()
             loss.backward()
             optimizer.step()
 
-            scores.append(score)
-            labels.append(label)
             losses.append(loss.item())
 
-        steps_in_epoch_trained = 0
-        if epoch % args.save_ckpt_by_epochs == 0:
-            save_checkpoint(model, logger, epoch, step, args)
+        model.eval()
+        batch = next(next(test_dataset_iter).iter_torch_batches(batch_size=1000))
+        loss, score, label = model(batch)
+        test_scores = [score]
+        test_labels = [label]
 
         session.report(
             {
```

Comment:
you probably want to put `epoch` for tracking as well..

Reply:
This is done automatically.

Reply:
To quote the docs: `It does not necessarily map to one epoch.`
Please update a for loop then:
```python
for epoch, epoch_pipe in enumerate(train_pipe.iter_epochs()):
    # to:
for epoch_pipe in train_pipe.iter_epochs():
```

Reply:
It does if you set the pipeline up how I did, but I will make this change.